include/zephyr/kernel.h

```diff
@@ -4574,6 +4575,30 @@ int k_msgq_cleanup(struct k_msgq *msgq);
  */
 __syscall int k_msgq_put(struct k_msgq *msgq, const void *data, k_timeout_t timeout);
 
+/**
+ * @brief Send a message to front of a message queue.
+ *
+ * This routine sends a message to message queue @a q, inserting it before any
+ * other messages already in the queue.
+ *
+ * @note The message content is copied from @a data into @a msgq and the @a data
```

Comment:
This should have a note or caveat clarifying the expected behaviour when the thread must pend because the message queue buffer is full as it is easy to miss/overlook.

If the message queue buffer is full, the calling thread may block. Only when it is unblocked can that message be inserted to the front of the message queue buffer. This means that the kernel can not guarantee LIFO ordering when the calling thread blocks.

Reply:
> This should have a note or caveat clarifying the expected behaviour when the thread must pend because the message queue buffer is full as it is easy to miss/overlook.
> 
> If the message queue buffer is full, the calling thread may block. Only when it is unblocked can that message be inserted to the front of the message queue buffer. This means that the kernel can not guarantee LIFO ordering when the calling thread blocks.

I think this scenario also guarantee LIFO because the calling thread prepend message to message queue.
And when new threads pending on a message queue, they also follows their sequence.

Reply:
LIFO can only be guaranteed within the message queue when there was space to add it to the message queue buffer. Consider the following ...

Let the message queue buffer have space for M messages. Let there be N threads already waiting to add messages into the message queue buffer. Let Thread X try to prepend a message but it is of lower priority than the N waiting threads; however it will be added to the end of the wait queue. If N >= M, then at least one of the threads ahead of Thread X in the wait queue will have its message added to and then retrieved from the message queue before Thread X can unpend have its message prepended. In this scenario, LIFO is not preserved.

Reply:
You are right. Maybe user need to know this information.